intrahost.py

```diff
@@ -129,157 +169,211 @@ def parser_tabfile_rename(parser=argparse.ArgumentParser()):
-                            a = a + consAllele[1:]
-                        assert a and a==a.upper()
-                        iSNVs[s][a] = f
-                    if util.misc.unique(map(len, iSNVs[s].keys())) == [1]:
-                        assert consAllele in iSNVs[s].keys()
-                elif s in consAlleles:
-                    # there is no iSNV data for this sample, so substitute the consensus allele
-                    iSNVs[s] = {consAlleles[s]:1.0}
-
-            # get unique allele list and map to numeric
-            alleles = [a for a,n in sorted(util.misc.histogram(consAlleles.values()).items(), key=lambda a,n:n, reverse=True) if a!=refAllele]
-            alleles2 = list(itertools.chain(*[iSNVs[s].keys() for s in samples if s in iSNVs]))
-            alleles = list(util.misc.unique([refAllele] + alleles + alleles2))
-            assert len(alleles)>1
-            alleleMap = dict((a,i) for i,a in enumerate(alleles))
-            genos = [str(alleleMap.get(consAlleles.get(s),'.')) for s in samples]
-            freqs = [(s in iSNVs) and ','.join(map(str, [iSNVs[s].get(a,0.0) for a in alleles[1:]])) or '.' for s in samples]
-
-            # prepare output row and write to file
-            out = [ref.id, pos, '.', alleles[0], ','.join(alleles[1:]), '.', '.', '.', 'GT:AF']
-            out = out + list(map(':'.join, zip(genos, freqs)))
-            outf.write('\t'.join(map(str, out))+'\n')
-
-def parser_vphaser_to_vcf(parser=argparse.ArgumentParser()):
-    parser.add_argument("inFile", help="Input vPhaser2 text file")
-    parser.add_argument("refFasta", help="Reference genome FASTA")
-    parser.add_argument("multiAlignment", help="Consensus genomes multi-alignment FASTA")
-    parser.add_argument("outVcf", help="Output VCF file")
+                            log.warn("dropping unclean consensus for %s at %s-%s: %s" % (s, pos, end, allele))
+                    
+                    # define genotypes and fractions
+                    iSNVs = {}
+                    for s in samples:
+                        
+                        # get all rows for this sample and merge allele counts together
+                        acounts = dict(itertools.chain.from_iterable(row['allele_counts']
+                            for row in rows if row['sample'] == s))
+                        if 'i' in acounts and 'd' in acounts:
+                            # this sample has both insertions and deletions at the same spot!
+                            # average the two (discordant) reference allele counts and
+                            # drop one of them (so we have only one reference allele
+                            acounts['i'] = int(round((acounts['i'] + acounts['d'])/2.0,0))
+                            del acounts['d']
+                        
+                        if acounts:
+                            # we have iSNV data on this sample
+                            consAllele = consAlleles[s]
```

Comment:
Based on your last comment, I just realized this assumption (that s in consAlleles) is not always right. It is true that iSNV calls can only exist where non-ambiguous consensus bases exist, so we'll never get an iSNV at a position that has an N. The problem is that there's logic upstream that expands the (pos,end) coordinates and make the variant bigger if necessary because of action happening in other samples. As such, an N might exist (or it might exceed the end of the consensus) and consAlleles might not have s.

How about we simply replace "if acounts" with "if acounts and s in consAlleles"?  If s is not in consAlleles, we shouldn't trust our ability to reconstruct the alleles for this sample, and the logic below should properly populate that sample with "no data" marks (the "." symbol).